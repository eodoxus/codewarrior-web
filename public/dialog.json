{
  "CrestfallenHome.CrestfallenMage": [
    "Hope...\nHope is forever lost...",
    "The Great Book of ECMAScript has been captured.\nAll I have left is this tattered page.",
    "It is said that a wizard in the north can release the book.\nIf only I could decipher the runes on this tattered page.\nPerhaps you can try?",
    "You've freed the Great Book!\nOh Hero, a great journey awaits you.\nIt is said the Great book can open many doors.\nGo east!"
  ]
}
