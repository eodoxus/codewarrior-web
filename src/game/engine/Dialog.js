import Url from "../../lib/Url";
import RestClient from "../../lib/RestClient";

export default class Dialog {
  static text;
  static async load() {
    Dialog.text = await new RestClient().get(Url.PUBLIC + "/dialog.json");
  }

  key;
  state;

  constructor(key, state) {
    this.key = key;
    this.state = state || 0;
  }

  getText() {
    const text = Dialog.text[this.key][this.state];
    return text;
  }

  getState() {
    return this.state;
  }

  setState(state) {
    this.state = state;
  }

  next() {
    const maxState = Dialog.text[this.key].length - 1;
    this.state = Math.min(maxState, this.state + 1);
  }
}
