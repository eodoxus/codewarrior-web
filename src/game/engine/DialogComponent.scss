$color-dialog--border: #fff;
$color-dialog--font: #fff;

.dialog {
  position: absolute;
  top: 5%;
  left: 20%;
  right: 20%;
  padding: 10px;
  min-height: 30px;
  border: 1px solid $color-dialog--border;
  border-radius: 4px;
  background: rgba(0, 0, 0, 0.7);
  color: $color-dialog--font;
}
