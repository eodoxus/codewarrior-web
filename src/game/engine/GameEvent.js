import Event from "../../lib/Event";

const listeners = [];

export default class GameEvent extends Event {
  static CANCEL = "cancel";
  static CLICK = "click";
  static CLOSE_TATTERED_PAGE = "closeTatteredPage";
  static CONFIRM = "confirm";
  static COLLISION = "collision";
  static DIALOG = "dialog";
  static DOORWAY = "doorway";
  static NPC_INTERACTION = "npcInteraction";
  static TALK = "talk";
  static TRANSITION = "transition";
  static OPEN_TATTERED_PAGE = "openTatteredPage";
  static STOP = "stop";

  static click(tile) {
    return new EventType(GameEvent.CLICK, tile);
  }

  static collision(entity) {
    return new EventType(GameEvent.COLLISION, entity);
  }

  static talk(entity) {
    return new EventType(GameEvent.TALK, entity);
  }

  static generic(eventName, data) {
    return new EventType(eventName, data);
  }

  static removeAllListeners() {
    listeners.forEach(listener => {
      listener.remove();
    });
  }

  static on(event, handler, enqueue = true) {
    const listener = super.on(event, handler);
    if (enqueue) {
      listeners.push(listener);
    }
    return listener;
  }
}

class EventType {
  type;
  data;
  constructor(type, data) {
    this.type = type;
    this.data = data;
  }

  getType() {
    return this.type;
  }

  getData() {
    return this.data;
  }
}
