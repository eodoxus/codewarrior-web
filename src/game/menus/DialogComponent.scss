@import "../../lib/lib.scss";

$color-dialog--border: #fff;
$color-dialog--font: #fff;

@mixin dialog-container {
  position: absolute;
  border: 1px solid $color-dialog--border;
  border-radius: 4px;
  background: rgba(0, 0, 0, 0.7);
}

.dialog {
  @include dialog-container;
  top: 5%;
  left: 20%;
  right: 20%;
  padding: 10px;
  min-height: 30px;
  color: $color-dialog--font;

  .confirm {
    @include dialog-container;
    padding: 10px;
    bottom: -67px;
    left: 50%;
    margin-left: -75px;

    button:first-child {
      margin-right: 10px;
    }
  }
}
