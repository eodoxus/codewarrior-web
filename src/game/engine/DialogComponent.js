import React, { Component } from "react";
import styles from "./DialogComponent.scss";
import Dialog from "./Dialog";
import GameEvent from "./GameEvent";

export default class DialogComponent extends Component {
  constructor(props) {
    super(props);
    this.state = { text: "" };
  }

  async componentDidMount() {
    await Dialog.load();
    GameEvent.on(GameEvent.DIALOG, this.onDialogEvent);
  }

  onDialogEvent = text => {
    this.setState({ text });
  };

  render() {
    if (!this.state.text) {
      return "";
    }
    const lines = this.state.text.split("\n").map((line, key) => {
      return (
        <span key={key}>
          {line}
          <br />
        </span>
      );
    });
    return <div className={styles.dialog}>{lines}</div>;
  }
}
