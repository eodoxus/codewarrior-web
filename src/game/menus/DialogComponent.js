import React, { Component } from "react";
import styles from "./DialogComponent.scss";
import cx from "classnames";
import Dialog from "../engine/Dialog";
import GameEvent from "../engine/GameEvent";
import { Button } from "../../components/forms/controls/buttons/index";

export default class DialogComponent extends Component {
  listener;

  constructor(props) {
    super(props);
    this.state = { dialog: "" };
  }

  async componentDidMount() {
    this.listener = GameEvent.on(GameEvent.DIALOG, this.onDialogEvent);
    await Dialog.load();
  }

  componentWillUnmount() {
    this.listener.remove();
    delete this.listener;
  }

  onCancel = e => {
    GameEvent.absorbClick(e);
    GameEvent.fire(GameEvent.CANCEL, e);
    this.setState({ dialog: "" });
  };

  onClick = e => {
    GameEvent.absorbClick(e);
    this.setState({ dialog: "" });
  };

  onConfirm = e => {
    GameEvent.absorbClick(e);
    GameEvent.fire(GameEvent.CONFIRM, { dialog: this.state.dialog });
    this.setState({ dialog: "" });
  };

  onDialogEvent = dialog => {
    this.setState({ dialog });
  };

  render() {
    if (!this.state.dialog) {
      return "";
    }

    const dialog = this.state.dialog;
    let text = "";
    let error = false;
    if (typeof dialog === "string") {
      text = dialog;
    } else if (dialog.confirm) {
      text = dialog.msg;
    } else if (dialog.error) {
      text = dialog.msg;
      error = true;
    }

    const lines = text
      .split("\n")
      .map((line, key) => <div key={key}>{line}</div>);

    return (
      <div
        className={cx(styles.dialog, error ? styles.error : "")}
        onClick={this.onClick}
      >
        <div className={styles.message}>{lines}</div>
        {this.renderConfirm(dialog)}
      </div>
    );
  }

  renderConfirm(dialog) {
    if (!dialog.confirm) {
      return;
    }
    return (
      <div className={styles.confirmContainer}>
        <div className={styles.confirm}>
          <Button text={dialog.confirm} onClick={this.onConfirm} />
          <Button text={dialog.cancel} onClick={this.onCancel} />
        </div>
      </div>
    );
  }
}
