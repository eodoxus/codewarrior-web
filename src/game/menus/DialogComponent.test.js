import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import Dialog from "../engine/Dialog";
import DialogComponent from "./DialogComponent";
import GameEvent from "../engine/GameEvent";

import dialog from "../../../public/dialog.json";

Dialog.load = () => {
  Dialog.text = dialog;
};

describe("<DialogComponent />", () => {
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<DialogComponent />, div);
    div.remove();
  });

  describe("Snapshots", () => {
    let node;
    beforeEach(() => {
      node = renderer.create(<DialogComponent />);
      const instance = node.getInstance();
    });

    afterEach(() => {
      GameEvent.removeAllListeners();
      node.unmount();
    });

    it("matches expected empty snapshot initially", () => {
      const tree = node.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("renders text dialog on dialog events", () => {
      GameEvent.fire(GameEvent.DIALOG, "test dialog");
      const tree = node.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("renders multiple divs for each line of dialog", () => {
      GameEvent.fire(GameEvent.DIALOG, "test dialog\nwith\nmultiple lines");
      const tree = node.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("renders a confirm dialog when confirm is present in dialog event", () => {
      GameEvent.fire(
        GameEvent.DIALOG,
        dialog["CrestfallenHome.CrestfallenMage"][2]
      );
      const tree = node.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("closes when dialog is empty", () => {
      GameEvent.fire(GameEvent.DIALOG, "test dialog");
      let tree = node.toJSON();
      expect(tree).toMatchSnapshot();

      GameEvent.fire(GameEvent.CLOSE_DIALOG);
      tree = node.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
