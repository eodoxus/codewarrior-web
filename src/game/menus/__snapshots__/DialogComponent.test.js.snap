// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DialogComponent /> Snapshots closes when dialog is empty 1`] = `
<div
  className={undefined}
>
  <div>
    test dialog
  </div>
</div>
`;

exports[`<DialogComponent /> Snapshots closes when dialog is empty 2`] = `""`;

exports[`<DialogComponent /> Snapshots matches expected empty snapshot initially 1`] = `""`;

exports[`<DialogComponent /> Snapshots renders a confirm dialog when confirm is present in dialog event 1`] = `
<div
  className={undefined}
>
  <div>
    It is said that a wizard in the north can release the book.
  </div>
  <div>
    If only I could decipher the runes on this tattered page.
  </div>
  <div>
    Perhaps you can try?
  </div>
  <div
    className={undefined}
  >
    <button
      className="btn btn-default btn-confirm"
      onClick={[Function]}
    >
      OK
    </button>
    <button
      className="btn btn-default btn-confirm"
      onClick={[Function]}
    >
      No thanks.
    </button>
  </div>
</div>
`;

exports[`<DialogComponent /> Snapshots renders multiple divs for each line of dialog 1`] = `
<div
  className={undefined}
>
  <div>
    test dialog
  </div>
  <div>
    with
  </div>
  <div>
    multiple lines
  </div>
</div>
`;

exports[`<DialogComponent /> Snapshots renders text dialog on dialog events 1`] = `
<div
  className={undefined}
>
  <div>
    test dialog
  </div>
</div>
`;
